src/Microsoft.OData.Edm/ExtensionMethods/EdmTypeSemantics.cs

```diff
@@ -1217,17 +1222,29 @@ public static EdmPrimitiveTypeKind PrimitiveKind(this IEdmTypeReference type)
         /// <returns>True if and only if the type inherits from the potential base type.</returns>
         public static bool InheritsFrom(this IEdmStructuredType type, IEdmStructuredType potentialBaseType)
         {
+            HashSet<string> lstTypes;
+            string fullName = type.FullTypeName();
+
+            if (typeDictionaryCache.TryGetValue(fullName, out lstTypes))
+            {
+                return lstTypes.Contains(potentialBaseType.FullTypeName());
+            }
+
+            lstTypes = new HashSet<string>();           
+
             do
             {
                 type = type.BaseType;
-                if (type != null && type.IsEquivalentTo(potentialBaseType))
+                if (type != null)
                 {
-                    return true;
+                    lstTypes.Add(type.FullTypeName());
                 }
             }
             while (type != null);
 
-            return false;
+            typeDictionaryCache.TryAdd(fullName, lstTypes);
+
+            return lstTypes.Contains(potentialBaseType.FullTypeName());

```

Comment:
I would factor out a "GetBaseTypes" static function that, given a type, returns a hashset of base types.  Then, in line 1229, if the value is not in the cache in line 1228, call this function.  Then, when you exit the check, you know you have the hashset (either from the cache or just created) and you can use it to see if the base type is in the list.